### PR TITLE
Use `HybridBitSet` in `SparseBitMatrix`.

### PR DESCRIPTION
**Pull request by nnethercote**
This fixes most of the remaining NLL memory regression.

r? @pnkfelix, because you reviewed #54286.
cc @nikomatsakis, because NLL
cc @Mark-Simulacrum, because this removes `array_vec.rs`
cc @lqd, because this massively improves `unic-ucd-name`, and probably other public crates

### CODE REVIEW

**Comment by nnethercote** [comment]
The submodules changes were accidental, and I have reverted them.

**Comment by nnethercote** [comment]
> [00:02:28] error: the lock file needs to be updated but --locked was passed to prevent this

The second commit is probably the cause of this, because it removes `array_vec.rs`... but I don't know what needs to be done.

**Comment by nnethercote** [comment]
@rust-timer build c243f30671409bce0ef56f65b318efac3c0012ff

**Comment by pnkfelix** [comment]
Looks great to me.

The dummy bitsets are unfortunate, but I'd rather land this first and then after that, maybe I will investigate revising the coding patterns to get rid of them.

**Comment by pnkfelix** [comment]
ah, I guess it would be prudent to wait until the results are back from rust-timer.

@bors r-

**Comment by nnethercote** [comment]
I expect the rust-timer results will be much the same as what I got locally, but sure, we can wait.

**Comment by lqd** [comment]
perf results are ready, awesome improvements for [max-rss](https://perf.rust-lang.org/compare.html?start=b80cb47889e0ad9400b6708ce2b4c4b364b71982&end=c243f30671409bce0ef56f65b318efac3c0012ff&stat=max-rss) as expected 🎉

File: src/librustc_data_structures/bit_set.rs [hunk 1]
```diff
@@ -341,21 +342,26 @@ impl<T: Idx> SparseBitSet<T> {
     }
 
     fn insert(&mut self, elem: T) -> bool {
-        // Ensure there are no duplicates.
-        if self.0.contains(&elem) {
-            false
+        assert!(self.len() < SPARSE_MAX);
+        if let Some(i) = self.0.iter().position(|&e| e >= elem) {
```
Comment (nikomatsakis) [inline src/librustc_data_structures/bit_set.rs]: I guess it doesn't seem worth using [`binary_search`](https://doc.rust-lang.org/std/primitive.slice.html#method.binary_search) for this?

Comment (nikomatsakis) [reply src/librustc_data_structures/bit_set.rs]: Nah, only `8` elements...

**Review by nikomatsakis — APPROVED** [review]

**Comment by nikomatsakis** [comment]
@bors p=1

This improvement is big enough to justify bumping the priority, I think. It will also help to make the RC2 candidate crater runs more successful.

**Comment by ljedrz** [comment]
Perf results are in; there are indeed substantial memory wins.

File: src/librustc_data_structures/bit_set.rs (hunk 1 above)
Comment (nnethercote) [reply src/librustc_data_structures/bit_set.rs]: Yeah, I haven't compared linear vs. binary, but I haven't seen evidence that this code is hot. I bet the difference is minimal, and if anything favours linear with that few elements.

**Comment by nnethercote** [comment]
> The dummy bitsets are unfortunate, but I'd rather land this first and then after that, maybe I will investigate revising the coding patterns to get rid of them.

I'd love to know how to avoid them. I asked about this for a similar example in [/r/rust](https://www.reddit.com/r/rust/comments/8ly3y8/need_help_with_veclike_container/) and the answers all involved multiple matches and `unreachable!()`s.

**Comment by nnethercote** [comment]
AFAICT, the `i686-pc-windows-gnu` machine hit the 3 hr limit. (The `x86_64-pc-windows-gnu` build took 2h58m.)

@bors retry, I guess

**Pull request merged**